### PR TITLE
feature: make trtllmsampler new_tokens format the universal format

### PR DESCRIPTION
**Pull request by netanel-haber**

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 1]
```diff
@@ -512,31 +512,56 @@ def create_py_executor_instance(
...
-    ):
-        sampler = get_spec_decoder(max_seq_len=model_engine.max_seq_len,
-                                   spec_config=model_engine.spec_config)
-    elif pytorch_backend_config.enable_trtllm_sampler:
+        return TorchStarAttentionSampler(max_seq_len=engine.max_seq_len)
+    trtllm_sampler = pytorch_backend_config.enable_trtllm_sampler
+    early_stop = not engine.model.model_config.is_generation
+    has_spec_dec = engine.spec_config is not None and engine.spec_config.spec_dec_mode.has_spec_decoder(
+    )
+    if sum([early_stop, trtllm_sampler, has_spec_dec]) > 1:
```
[the remark this answers is not in the document]
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/_util.py:545]: Also please rename variable `trtllm_sampler` to `is_trtllm_sampler` or `enable_trtllm_sampler`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -1216,15 +1213,15 @@ def _prepare_tp_inputs(
...
+        # will contain previous batch indices of generation requests
         previous_batch_indices = []
         previous_pos_indices = []
         for request in extend_requests:
             # the request has no previous tensor:
             # (1) next_draft_tokens_device is None, which means overlap scheduler is disabled; or
             # (2) a dummy request; or
             # (3) the first step in the generation server of disaggregated serving
-            if next_draft_tokens_device is None or request.is_dummy or request.py_batch_idx is None:
+            if not is_mtp or request.is_dummy or request.py_batch_idx is None:
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1224]: Should this take into account Eagle-3 as well?

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (netanel-haber) [reply tensorrt_llm/_torch/pyexecutor/_util.py:545]: Can TRTLLMSampler handle an EarlyStop? If so you're right, and I have to take that into account

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 3]
```diff
@@ -34,16 +34,11 @@ def values(self):
...
+@dataclass(kw_only=True, frozen=True)
 class SampleState:
     scheduled_requests: ScheduledRequests
 
-    logits: torch.Tensor = None
-    logits_host: torch.Tensor = None
-
-    # Set when decode_async() has evaluated these to avoid computing again in update_requests()
-    # log_probs[request_idx][token_idx]
-    log_probs: list[list[float] | None] | None = None
```
[the remark this answers is not in the document]
Comment (netanel-haber) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:46]: Nope, just wanted to be precise to the state of the code ATM - but I can revert this change

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/_util.py:546]: Yes, `is_generation` should be used in disagg serving, and it was supported by TRTLLMSampler last time I tried. And yes, it supports early stopping too.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -1196,28 +1193,26 @@ def _prepare_tp_inputs(
...
         num_ctx_tokens = len(input_ids)
         new_tokens_device, new_tokens_lens_device, next_draft_tokens_device = None, None, None
         if new_tensors_device is not None:
             # speculative decoding cases: [batch, 1 + draft_len], others: [batch]
             new_tokens_device = new_tensors_device.new_tokens
-            if self.without_logits:
-                assert isinstance(new_tensors_device, SampleStateTensorsMTP)
-                new_tokens_lens_device = new_tensors_device.new_tokens_lens  # [batch]
-                next_draft_tokens_device = new_tensors_device.next_draft_tokens  # [batch, draft_len]
+        is_mtp = isinstance(new_tensors_device, SampleStateTensorsMTP)
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1204]: Shouldn't this code go inside the `if new_tensors_device is not None:` block?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 2 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1245]: In that case, `is_mtp` is not a good name, as Eagle-3 and MTP are different things. Perhaps `is_spec_decoding`?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -1336,12 +1329,21 @@ def _prepare_tp_inputs(
             prompt_lengths.append(request.py_prompt_len)
             draft_lens.append(0)
 
-            request.py_batch_idx = batch_idx
-            batch_idx += 1
+            request.py_batch_idx = request.seq_slot
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1332]: Why do we need to store `seq_slot` on `py_batch_idx`? Under what circumstance do these two diverge?

File: tensorrt_llm/_torch/pyexecutor/scheduler.py [hunk 6]
```diff
@@ -36,9 +36,8 @@ def can_run_cuda_graph(self) -> bool:
     def batch_size(self) -> int:
         return len(self.context_requests) + len(self.generation_requests)
 
-    @property
-    def all_requests(self) -> chain[LlmRequest]:
-        return chain(self.context_requests, self.generation_requests)
+    def all_requests(self) -> list[LlmRequest]:
+        return list(chain(self.context_requests, self.generation_requests))
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/scheduler.py:40]:
```suggestion
    def all_requests(self) -> list[LlmRequest]:
        return self.context_requests + self.generation_requests
```

Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/scheduler.py:40]:
```suggestion
    def all_requests(self) -> list[LlmRequest]:
        return self.context_requests + self.generation_requests
```

Comment (netanel-haber) [reply tensorrt_llm/_torch/pyexecutor/scheduler.py:40]: When Python is nice, it's nice
```python
>>> a = [1]
>>> b = [2]
>>> c = a + b
>>> a[0] = 5
>>> c
[1, 2]
```

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 5 above)
Comment (netanel-haber) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1332]: That was the question I asked you way back when - my rationale is that a request may be evicted between between batch runs and lose its seq_slot etc. - I couldn't determine whether or not this was true or not, so I played it safe.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 4 above)
Comment (netanel-haber) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1204]: No, not necessarily, since `new_tokens_lens_device` and `next_draft_tokens_device` are required to exist if `isinstance(new_tensors_device, SampleStateTensorsMTP)`  - basically, this is a more explicit typeguard

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 5 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1333]: I'm pretty sure now that overlap scheduling and evicting requests do not work together, given that there is a requirement that the last step previous tensors device be passed [on the next iteration](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/py_executor.py#L1067-L1068). Therefore, a current restriction of the PyTorch framework is that overlap scheduling be scheduled with GuaranteedNoEvict CC @Funatiq @QiJune .

With that restriction in mind, I think it is safe to remove `request.py_batch_idx` and just use `seq_slot`.

**Review by dcampora — APPROVED** [review]

**Review by DomBrown — APPROVED** [review]

**Review by suyoggupta — APPROVED** [review]
AD changes LGTM
